File: bindings/python/examples/topp3_socp.py
"""Use TOPP3-SOCP to convert a JAX-defined path into a jerk-limited time-optimal trajectory.

The example builds a TOPP2-RA seed, solves a third-order ``(a,b)`` profile with
Clarabel, and performs one refinement around the first solution.
"""

import numpy as np

import copp_py as copp


def main() -> None:
    try:
        import jax
        import jax.numpy as jnp
    except ImportError as exc:
        raise SystemExit(
            'Install JAX to run this example: python -m pip install "copp-py[jax]"'
        ) from exc

    jax.config.update("jax_enable_x64", True)

    dim = 3
    n = 1001
    dt = 1.0e-3

    # 1) Define q(s). Path.from_jax differentiates it up to third order.
    def q_fn(s):
        freq = jnp.array([2.0 * jnp.pi, 3.0 * jnp.pi, 5.0 * jnp.pi], dtype=jnp.float64)
        phase = jnp.array([0.0, 0.3, 0.7], dtype=jnp.float64)
        return jnp.sin(freq * s + phase)

    path = copp.Path.from_jax(q_fn, 0.0, 1.0)
    s = np.linspace(0.0, 1.0, n, dtype=np.float64)

    # 2) Build robot constraints (3-axis), then apply symmetric limits
    # velocity/acceleration/jerk = [-1, 1].
    robot = copp.Robot(dim, capacity=n)
    robot.append_s(s)
    robot.set_q_from_path_3rd(path, 0, n)

    upper = np.ones(dim, dtype=np.float64)
    lower = -upper
    robot.add_velocity_limits(upper, lower, start_idx_s=0, length=n)
    robot.add_acceleration_limits(upper, lower, start_idx_s=0, length=n)
    robot.add_jerk_limits(upper, lower, start_idx_s=0, length=n)

    # 3) Use TOPP2-RA as the initial a(s) linearization for the third-order model.
    topp2_problem = copp.solver.topp2_ra.Problem(
        robot.constraints,
        idx_s_interval=(0, n - 1),
        a_boundary=(0.0, 0.0),
    )
    a_ra0 = copp.solver.topp2_ra.solve(topp2_problem, copp.solver.topp2_ra.Options())
    robot.constraints.amax_substitute(a_ra0, 0)

    # 4) Solve TOPP3-SOCP twice, refreshing the a(s) linearization once.
    options = copp.solver.topp3_socp.Options(allow_almost_solved=True)

    problem1 = copp.solver.topp3_socp.Problem(
        robot.constraints,
        a_ra0,
        idx_s_start=0,
        a_boundary=(0.0, 0.0),
        b_boundary=(0.0, 0.0),
        num_stationary_max=1,
    )
    profile1 = copp.solver.topp3_socp.solve(problem1, options)
    t_final1, t_s1 = copp.interpolation.s_to_t_topp3(s, profile1, 0.0)
    s_t1 = copp.interpolation.t_to_s_topp3_uniform(
        s,
        profile1,
        t_s1,
        dt,
        t0=0.0,
        include_final=True,
    )

    problem2 = copp.solver.topp3_socp.Problem(
        robot.constraints,
        profile1.a,
        idx_s_start=0,
        a_boundary=(0.0, 0.0),
        b_boundary=(0.0, 0.0),
        num_stationary_max=1,
    )
    profile2 = copp.solver.topp3_socp.solve(problem2, options)
    t_final2, t_s2 = copp.interpolation.s_to_t_topp3(s, profile2, 0.0)
    s_t2 = copp.interpolation.t_to_s_topp3_uniform(
        s,
        profile2,
        t_s2,
        dt,
        t0=0.0,
        include_final=True,
    )

    # 5) Print the tutorial summary.
    print("TOPP3-SOCP done. (The first iteration)")
    print(f"dim = {dim}, N = {n}")
    print(f"t_final = {t_final1:.6f} s")
    print(f"a_profile.len() = {len(profile1.a)}")
    print(f"b_profile.len() = {len(profile1.b)}")
    print(f"s(t) samples = {len(s_t1)}")
    print("---------")
    print("TOPP3-SOCP done. (The second iteration)")
    print(f"t_final = {t_final2:.6f} s <= {t_final1:.6f} s")
    print(f"a_profile.len() = {len(profile2.a)}")
    print(f"b_profile.len() = {len(profile2.b)}")
    print(f"s(t) samples = {len(s_t2)}")


if __name__ == "__main__":
    main()

File: bindings/cpp/examples/topp3_socp.cpp
// Tutorial example: solve TOPP3-SOCP with two SCP refinement rounds.
//
// The SOCP formulation depends on a linearization profile. A common workflow is
// to solve once with a simple initial `a0`, then rebuild the problem with the
// recovered `profile.a` and solve again.

#include <iostream>
#include <vector>

#include <copp/copp.hpp>

int main()
{
    // Step 1: Minimal one-dimensional constraint setup.
    std::vector<double> s{0.0, 0.5, 1.0};
    std::vector<double> amax{1.0, 1.0, 1.0};
    std::vector<double> a0{0.25, 0.25, 0.25};

    copp::Constraints constraints(1, s.size());
    constraints.append_s(s)
        .add_constraint_1st(amax, 0);

    // Step 2: Configure Clarabel and endpoint boundary values.
    const copp::Boundary3 boundary{0.25, 0.25, 0.0, 0.0};

    copp::clarabel::Options options;
    options.allow_almost_solved = true;

    // Step 3: First SCP round using the hand-chosen linearization profile.
    copp::solver::topp3::Problem first_problem{
        constraints.ref(),
        a0,
        0,
        boundary,
    };
    auto profile1 = copp::solver::topp3_socp::solve(first_problem, options);

    // Step 4: Second SCP round using the previous `a` profile as the new
    // linearization. More rounds can be added by repeating this pattern.
    copp::solver::topp3::Problem second_problem{
        constraints.ref(),
        profile1.a,
        0,
        boundary,
    };
    auto profile2 = copp::solver::topp3_socp::solve(second_problem, options);

    // Step 5: Interpolate the final third-order profile in time.
    auto time = copp::interpolation::s_to_t_topp3(s, profile2, 0.0);

    std::cout << "TOPP3-SOCP done with 2 refinement rounds.\n";
    std::cout << "profile length = " << profile2.len() << "\n";
    std::cout << "t_final = " << time.t_final << "\n";

    return 0;
}
